l1-contracts/contracts/bridge/L1NativeTokenVault.sol

```diff
@@ -0,0 +1,161 @@
+    function bridgeBurn(
+        uint256 _chainId,
+        uint256 _mintValue,
+        bytes32 _assetInfo,
+        address _prevMsgSender,
+        bytes calldata _data
+    ) external payable override onlyBridge whenNotPaused returns (bytes memory _bridgeMintData) {
+        uint256 _depositAmount = abi.decode(_data, (uint256));
+
+        uint256 amount;
+        address l1Token = tokenAddress[_assetInfo];
+        if (l1Token == ETH_TOKEN_ADDRESS) {
+            amount = msg.value;
+            require(_depositAmount == 0, "L1SharedBridge: msg.value not equal to amount");
+        } else {
+            // The Bridgehub also checks this, but we want to be sure
+            require(msg.value == 0, "ShB m.v > 0 b d.it");
+            amount = _depositAmount;
+
+            uint256 withdrawAmount = _depositFunds(_prevMsgSender, IERC20(l1Token), _depositAmount); // note if _prevMsgSender is this contract, this will return 0. This does not happen.
+            require(withdrawAmount == _depositAmount, "3T"); // The token has non-standard transfer logic
+        }
+        require(amount != 0, "6T"); // empty deposit amount
+
+        if (L1_SHARED_BRIDGE.hyperbridgingEnabled(_chainId)) {
+            chainBalance[_chainId][l1Token] += _depositAmount;
+        }
+
+        _bridgeMintData = abi.encode(amount, _getERC20Getters(l1Token)); // to do add l2Receiver in here
+    }
+
+    /// @dev Transfers tokens from the depositor address to the smart contract address.
+    /// @return The difference between the contract balance before and after the transferring of funds.
+    function _depositFunds(address _from, IERC20 _token, uint256 _amount) internal returns (uint256) {
+        uint256 balanceBefore = _token.balanceOf(address(this));
+        // slither-disable-next-line arbitrary-send-erc20
+        _token.safeTransferFrom(_from, address(this), _amount);
+        uint256 balanceAfter = _token.balanceOf(address(this));
+
+        return balanceAfter - balanceBefore;
+    }
+
+    /// @dev Receives and parses (name, symbol, decimals) from the token contract
+    function _getERC20Getters(address _token) internal view returns (bytes memory) {
+        if (_token == ETH_TOKEN_ADDRESS) {
+            bytes memory name = bytes("Ether");
+            bytes memory symbol = bytes("ETH");
+            bytes memory decimals = abi.encode(uint8(18));
+            return abi.encode(name, symbol, decimals); // when depositing eth to a non-eth based chain it is an ERC20
+        }
+
+        (, bytes memory data1) = _token.staticcall(abi.encodeCall(IERC20Metadata.name, ()));
+        (, bytes memory data2) = _token.staticcall(abi.encodeCall(IERC20Metadata.symbol, ()));
+        (, bytes memory data3) = _token.staticcall(abi.encodeCall(IERC20Metadata.decimals, ()));
+        return abi.encode(data1, data2, data3);
+    }
+
+    function bridgeMint(address _account, uint256 _amount) external payable override {}
```

Comment:
Shall we add finalizeWithdrawal logic here?

I can see that we moved the deposit, but withdraw is still in L1SharedBridge

Reply:
Yes, I'm in the process of moving the things from L1SharedBridge, I will push